File: src/apple/Packages/MapleUI/Sources/MapleUI/Atoms/MuiBadge.swift
// MuiBadge.swift — Maple UI Badge atom.
// Contract: docs/design/maple-ui/components/badge.md

import SwiftUI

public enum MuiBadgeVariant: Sendable {
    case count, signal, rating
}

/// A small, filled status/count indicator. Informational, not interactive —
/// badges carry no hover/pressed/focused state (badge.md §States).
public struct MuiBadge: View {
    public let variant: MuiBadgeVariant
    public let value: String
    /// Accessible description when `value` alone isn't self-explanatory
    /// (e.g. "3 unread" rather than just "3") (badge.md §Props).
    public let label: String?

    public init(variant: MuiBadgeVariant = .count, value: String, label: String? = nil) {
        self.variant = variant
        self.value = value
        self.label = label
    }

    public var body: some View {
        Group {
            switch variant {
            case .count:
                pill(background: MuiTokens.primaryDim, foreground: MuiTokens.textMain)
            case .signal:
                pill(background: MuiTokens.warn.opacity(0.25), foreground: MuiTokens.textMain)
            case .rating:
                HStack(spacing: 2) {
                    MuiIcon(name: "star.fill", size: .xs, color: MuiTokens.star)
                    Text(value)
                        .font(MuiTokens.TypeScale.font(.chipLabel))
                        .foregroundStyle(MuiTokens.textMain)
                }
            }
        }
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(Self.accessibleLabel(variant: variant, value: value, label: label))
    }

    @ViewBuilder
    private func pill(background: Color, foreground: Color) -> some View {
        Text(value)
            .font(MuiTokens.TypeScale.font(.chipLabel))
            .foregroundStyle(foreground)
            .padding(.horizontal, MuiTokens.spacingXs)
            .padding(.vertical, 2)
            .background(background, in: Capsule())
    }

    /// The accessible label a badge exposes to assistive technology — an
    /// explicit `label` override when supplied, otherwise `value` itself,
    /// except `rating` which is announced as "Rating N" so the numeric
    /// value reads correctly rather than just the glyph count (badge.md
    /// §Accessibility). Public + static so this logic is unit-testable
    /// without rendering a view.
    public static func accessibleLabel(variant: MuiBadgeVariant, value: String, label: String?) -> String {
        if let label { return label }
        return variant == .rating ? "Rating \(value)" : value
    }
}

#Preview("MuiBadge — Variants") {
    HStack(spacing: 12) {
        MuiBadge(variant: .count, value: "3")
        MuiBadge(variant: .signal, value: "Needs review")
        MuiBadge(variant: .rating, value: "4")
    }
    .padding()
    .background(MuiTokens.bg)
}

#Preview("MuiBadge — With accessible label") {
    MuiBadge(variant: .count, value: "3", label: "3 unread")
        .padding()
        .background(MuiTokens.bg)
}

File: src/windows/Maple.WinUI/MapleUI/Atoms/MuiBadge.cs
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Automation;
using Microsoft.UI.Xaml.Controls;
using Microsoft.UI.Xaml.Media;
using Windows.UI;

namespace Maple.UI.Atoms
{
    /// <summary>Badge variant (badge.md § Variants).</summary>
    public enum MuiBadgeVariant { Count, Signal, Rating }

    /// <summary>
    /// Maple.UI Badge atom (docs/design/maple-ui/components/badge.md) — a
    /// small, filled, non-interactive status/count indicator.
    /// </summary>
    public sealed class MuiBadge : ContentControl
    {
        public static readonly DependencyProperty VariantProperty =
            DependencyProperty.Register(nameof(Variant), typeof(MuiBadgeVariant), typeof(MuiBadge),
                new PropertyMetadata(MuiBadgeVariant.Count, (d, _) => ((MuiBadge)d).Rebuild()));

        public static readonly DependencyProperty ValueProperty =
            DependencyProperty.Register(nameof(Value), typeof(string), typeof(MuiBadge),
                new PropertyMetadata(string.Empty, (d, _) => ((MuiBadge)d).Rebuild()));

        public static readonly DependencyProperty LabelProperty =
            DependencyProperty.Register(nameof(Label), typeof(string), typeof(MuiBadge),
                new PropertyMetadata(null, (d, _) => ((MuiBadge)d).Rebuild()));

        public MuiBadgeVariant Variant
        {
            get => (MuiBadgeVariant)GetValue(VariantProperty);
            set => SetValue(VariantProperty, value);
        }

        /// <summary>The displayed content — a count, a rating number, or
        /// short signal text.</summary>
        public string Value
        {
            get => (string)GetValue(ValueProperty);
            set => SetValue(ValueProperty, value);
        }

        /// <summary>Accessible description when Value alone isn't
        /// self-explanatory (badge.md: e.g. "3 unread" rather than "3").
        /// Falls back to Value when unset.</summary>
        public string? Label
        {
            get => (string?)GetValue(LabelProperty);
            set => SetValue(LabelProperty, value);
        }

        private readonly Border _pill = new() { CornerRadius = new CornerRadius(999) };
        private readonly TextBlock _pillText = new() { FontSize = 11, FontWeight = Microsoft.UI.Text.FontWeights.Bold };
        private readonly StackPanel _ratingRow = new() { Orientation = Orientation.Horizontal, Spacing = 4 };
        private readonly MuiIcon _star = new() { IconName = "star-filled", Size = MuiIconSize.Xs14 };
        private readonly TextBlock _ratingText = new() { FontSize = 11 };

        public MuiBadge()
        {
            _pill.Child = _pillText;
            _ratingRow.Children.Add(_star);
            _ratingRow.Children.Add(_ratingText);
            IsTabStop = false;
            AutomationProperties.SetAccessibilityView(this, Microsoft.UI.Xaml.Automation.Peers.AccessibilityView.Content);
            Rebuild();
        }

        private static Brush R(string key) => (Brush)Application.Current.Resources[key];

        private void Rebuild()
        {
            _pillText.Text = Value ?? string.Empty;
            _pill.Padding = new Thickness(6, 2, 6, 2);

            switch (Variant)
            {
                case MuiBadgeVariant.Signal:
                    var warn = ((SolidColorBrush)R("MapleWarn")).Color;
                    _pill.Background = new SolidColorBrush(Color.FromArgb(0x40, warn.R, warn.G, warn.B));
                    _pill.BorderBrush = R("MapleWarn");
                    _pill.BorderThickness = new Thickness(1);
                    _pillText.Foreground = R("MapleWarn");
                    Content = _pill;
                    break;

                case MuiBadgeVariant.Rating:
                    _ratingText.Text = Value ?? string.Empty;
                    _ratingText.Visibility = string.IsNullOrEmpty(Value) ? Visibility.Collapsed : Visibility.Visible;
                    _ratingText.Foreground = R("MapleStar");
                    Content = _ratingRow;
                    break;

                default: // Count
                    _pill.Background = R("MaplePrimaryDim");
                    _pill.BorderBrush = null;
                    _pill.BorderThickness = new Thickness(0);
                    _pillText.Foreground = R("MapleTextMain");
                    Content = _pill;
                    break;
            }

            // badge.md § Accessibility: never convey meaning by color alone —
            // always expose a name, falling back to the raw Value.
            var name = Label ?? Value ?? string.Empty;
            if (!string.IsNullOrEmpty(name))
                AutomationProperties.SetName(this, name);
        }
    }
}
